File: java/Shell.java
public class Shell {
	public static void sort(int a[]) {
		int h = 0;
		while(h < a.length)
			h = 3*h+1;
		while(h > 0) {
			h = h/3;
			for(int k = 0; k < h; k++) {
				for(int j = 0; j < a.length/h; j += a.length/h)
					insertion(a, j, a.length/h-j);
			}
		}
	}

	private static void insertion(int a[], int l, int n) {
		for(int i = l+2; i < l+n; i++) {
			for(int k = i; k > l && a[k] < a[k-1]; k--)
				Util.swap(a, k, k-1);
		}
	}
}

File: php/shell.php
<?php
require_once('util.php');

// shell implements shell sort algorithm. The function takes an array.
function shell(&$a) {
	$h = 0;
	while($h < count($a))
		$h = 3*$h+1;
	while($h > 0) {
		$h = floor($h/3);
		for($k = 0; $k < $h; $k++) {
			for($j = 0; $j < floor(count($a)/$h); $j += floor(count($a)/$h))
				_insertion($a, $j, floor(count($a)/$h)-$j);
		}
	}
}

function _insertion(&$a, $l, $n) {
	for($i = $l+2; $i < $l+$n; $i++) {
		for($k = $i; $k > $l && $a[$k] < $a[$k-1]; $k--)
			swap($a, $k, $k-1);
	}
}
